applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp

```diff
@@ -111,224 +129,531 @@ Element::Pointer NodalConcentratedElement::Clone(
+        rValues.resize( system_size, false );
+
+    // Auxiliar index
+    IndexType aux_index = 0;
+
+    // The displacement terms
+    if( mELementalFlags.Is(NodalConcentratedElement::COMPUTE_DISPLACEMENT_STIFFNESS) ||
+        mELementalFlags.Is(NodalConcentratedElement::COMPUTE_NODAL_MASS)) {
 
-    if ( rValues.size() != dimension )
-        rValues.resize( dimension, false );
+        rValues[0] = GetGeometry()[0].GetSolutionStepValue( DISPLACEMENT_X, Step );
+        rValues[1] = GetGeometry()[0].GetSolutionStepValue( DISPLACEMENT_Y, Step );
 
-    rValues[0] = GetGeometry()[0].GetSolutionStepValue( DISPLACEMENT_X, Step );
-    rValues[1] = GetGeometry()[0].GetSolutionStepValue( DISPLACEMENT_Y, Step );
+        if ( dimension == 3 )
+            rValues[2] = GetGeometry()[0].GetSolutionStepValue( DISPLACEMENT_Z, Step );
 
-    if ( dimension == 3 )
-        rValues[2] = GetGeometry()[0].GetSolutionStepValue( DISPLACEMENT_Z, Step );
+        aux_index += dimension;
+    }
+
+    // The rotational terms
+    if( mELementalFlags.Is(NodalConcentratedElement::COMPUTE_ROTATIONAL_STIFFNESS) ||
+        mELementalFlags.Is(NodalConcentratedElement::COMPUTE_NODAL_INERTIA)) {
+
+        rValues[aux_index + 0] = GetGeometry()[0].GetSolutionStepValue( ROTATION_X, Step );
+        rValues[aux_index + 1] = GetGeometry()[0].GetSolutionStepValue( ROTATION_Y, Step );
+
+        if ( dimension == 3 )
+            rValues[aux_index + 2] = GetGeometry()[0].GetSolutionStepValue( ROTATION_Z, Step );
+    }
 }
 
 
 //************************************VELOCITY****************************************
-//************************************************************************************
+/***********************************************************************************/
 
 void NodalConcentratedElement::GetFirstDerivativesVector( Vector& rValues, int Step )
 {
     //GIVES THE VECTOR WITH THE TIME DERIVATIVE OF THE DOFS VARIABLES OF THE ELEMENT (i.e. ELEMENT VELOCITIES)
-    const unsigned int dimension = GetGeometry().WorkingSpaceDimension();
+    const SizeType dimension = GetGeometry().WorkingSpaceDimension();
+
+    // Resizing as needed
+    const SizeType system_size = ComputeSizeOfSystem();
 
-    if ( rValues.size() != dimension )
-        rValues.resize( dimension, false );
+    if ( rValues.size() != system_size )
+        rValues.resize( system_size, false );
+
+    // Auxiliar index
+    IndexType aux_index = 0;
+
+    // The displacement terms
+    if( mELementalFlags.Is(NodalConcentratedElement::COMPUTE_DISPLACEMENT_STIFFNESS) ||
+        mELementalFlags.Is(NodalConcentratedElement::COMPUTE_NODAL_MASS)) {
+
+        rValues[0] = GetGeometry()[0].GetSolutionStepValue( VELOCITY_X, Step );
+        rValues[1] = GetGeometry()[0].GetSolutionStepValue( VELOCITY_Y, Step );
+
+        if ( dimension == 3 )
+            rValues[2] = GetGeometry()[0].GetSolutionStepValue( VELOCITY_Z, Step );
+
+        aux_index += dimension;
```

Comment:
why do you need the aux index here?

Reply:
To track when adding displacement/rotation with any combination possible